ams/ams-server/src/main/java/com/netease/arctic/ams/server/service/impl/TableExpireService.java

```diff
@@ -103,69 +101,94 @@ public void run() {
+        if (baseTable == null) {
+          LOG.warn("[{}] Base table is null: {} ", traceId, tableIdentifier);
+          return null;
         }
-      } catch (Throwable t) {
-        LOG.error("[" + traceId + "] unexpected expire error of table " + tableIdentifier, t);
+        UnkeyedTable changeTable = keyedArcticTable.changeTable();
+        if (changeTable == null) {
+          LOG.warn("[{}] Change table is null: {}", traceId, tableIdentifier);
+          return null;
+        }
+
+        // get valid files in the change store which shouldn't physically delete when expire the snapshot
+        // in the base store
+        Set<String> baseExclude = UnKeyedTableUtil.getAllContentFilePath(changeTable);
+        baseExclude.addAll(finalHiveLocation);
+        long latestBaseFlinkCommitTime = fetchLatestFlinkCommittedSnapshotTime(baseTable);
+        expireSnapshots(baseTable, Math.min(latestBaseFlinkCommitTime, startTime - baseSnapshotsKeepTime),
+            baseExclude);
+        long baseCleanedTime = System.currentTimeMillis();
+        LOG.info("[{}] {} base expire cost {} ms", traceId, arcticTable.id(), baseCleanedTime - startTime);
+
+        // delete ttl files
+        List<DataFileInfo> changeDataFiles = ServiceContainer.getFileInfoCacheService()
+            .getChangeTableTTLDataFiles(keyedArcticTable.id().buildTableIdentifier(),
+                System.currentTimeMillis() - changeDataTTL);
+        deleteChangeFile(keyedArcticTable, changeDataFiles);
+
+        // get valid files in the base store which shouldn't physically delete when expire the snapshot
+        // in the change store
+        Set<String> changeExclude = UnKeyedTableUtil.getAllContentFilePath(baseTable);
+        changeExclude.addAll(finalHiveLocation);
+        long latestChangeFlinkCommitTime = fetchLatestFlinkCommittedSnapshotTime(changeTable);
+        expireSnapshots(changeTable, Math.min(latestChangeFlinkCommitTime, startTime - changeSnapshotsKeepTime),
+            changeExclude);
+        return null;
+      });
+      LOG.info("[{}] {} expire cost total {} ms", traceId, arcticTable.id(),
+          System.currentTimeMillis() - startTime);
+    } else {
+      UnkeyedTable unKeyedArcticTable = arcticTable.asUnkeyedTable();
+      long latestFlinkCommitTime = fetchLatestFlinkCommittedSnapshotTime(unKeyedArcticTable);
+      expireSnapshots(unKeyedArcticTable, Math.min(latestFlinkCommitTime, startTime - baseSnapshotsKeepTime),
+          hiveLocation);
+      long baseCleanedTime = System.currentTimeMillis();
+      LOG.info("[{}] {} unKeyedTable expire cost {} ms", traceId, arcticTable.id(), baseCleanedTime - startTime);
+    }
+  }
+
+  public static long fetchLatestFlinkCommittedSnapshotTime(UnkeyedTable table) {
+    long latestCommitTime = Long.MAX_VALUE;
+    for (Snapshot snapshot : table.snapshots()) {
+      if (snapshot.summary().containsKey("flink.max-committed-checkpoint-id")) {
```

Comment:
I suggest using a static variable to hold this value and link it to the iceberg flink writer class in the comment.

Reply:
I import a static variable for this constant and add some comment linking to `org.apache.iceberg.flink.sink.IcebergFilesCommitter#MAX_COMMITTED_CHECKPOINT_ID`, but the `{@link xxx}` doesn't work here because it is inaccessible.